### PR TITLE
continue to load candidate device when load device failed

### PR DESCRIPTION
**Pull request by tiger100256-hu**
### Details:
 - *continue to load candidate device when load device failed *
 - *wait for the first ready device instead of CPU*
 - *fix potential risk: the containers should be initialized totally be run tasks in threads*

### Tickets:
 - CVS-59952

### CODE REVIEW

**Comment by myshevts** [comment]
we need to invent how we test possible combinations of failures with behavior tests....

**Comment by myshevts** [comment]
overall the logic seems to be valid, but implemented so complex way, that some corner cases may slip thru... pls check if it is possible to make it simpler....

File: inference-engine/src/multi_device/multi_device_exec_network.cpp [hunk 1]
```diff
@@ -193,55 +193,123 @@ MultiDeviceExecutableNetwork::MultiDeviceExecutableNetwork(const std::string&
...
         _workerRequests[p.deviceName];
         _inferPipelineTasksDeviceSpecific[p.deviceName] = NULL;
-        const auto device = p.deviceName;
-        const auto deviceConfig = p.config;
+        auto device = p.deviceName;
+        auto deviceConfig = p.config;
         // will not wait for loading accelerator network,
         // so some parameters need to be transferred by value.
-       _executor->run([&, modelPath, network, device, deviceConfig]() {
+        loads.push_back([&, modelPath, network, device, deviceConfig,
```
[the remark this answers is not in the document]
Comment (tiger100256-hu) [reply inference-engine/src/multi_device/multi_device_exec_network.cpp]: fix it, use "this" replace "&"

File: inference-engine/src/multi_device/multi_device_exec_network.cpp [hunk 2]
```diff
@@ -193,55 +193,123 @@ MultiDeviceExecutableNetwork::MultiDeviceExecutableNetwork(const std::string&
...
+            do {
+                try {
+                    if (!modelPath.empty()) {
+                        executableNetwork = _core->LoadNetwork(modelPath, device, deviceConfig);
+                    } else {
+                        executableNetwork = _core->LoadNetwork(network, device, deviceConfig);
+                    }
+                    loadSuccess = true;
+                } catch (const std::exception& e) {
+                    printf("Warning: load network to %s failed: %s\n", device.c_str(), e.what());
```
[the remark this answers is not in the document]
Comment (tiger100256-hu) [reply inference-engine/src/multi_device/multi_device_exec_network.cpp]: fix it, save the failed message into a member of load context, can print it in Debug log if need

File: inference-engine/src/multi_device/multi_device_exec_network.cpp [hunk 3]
```diff
@@ -193,55 +193,123 @@ MultiDeviceExecutableNetwork::MultiDeviceExecutableNetwork(const std::string&
...
+                // load failed and device is not cpu, try to select next device
+                if (!loadSuccess && !curDevIsCPU) {
+                    auto eraseDevice = std::find_if(deviceList.begin(), deviceList.end(),
+                            [device](DeviceInformation& d){
+                                return d.deviceName == device;
+                            });
+                    deviceList.erase(eraseDevice);
+                    try {
+                        if (!deviceList.empty()) {
+                            _acceleratorDevice = _multiPlugin->SelectDevice(deviceList, networkPrecision);
```
[the remark this answers is not in the document]
Comment (tiger100256-hu) [reply inference-engine/src/multi_device/multi_device_exec_network.cpp]: the load  logic is that select a device from device list , if the device load failed, then remove the device from device list and then select and load again. The loop is not finished until load success or there is no device list in device list or the select function throw exception

File: inference-engine/src/multi_device/multi_device_exec_network.cpp [hunk 4]
```diff
@@ -193,55 +193,123 @@ MultiDeviceExecutableNetwork::MultiDeviceExecutableNetwork(const std::string&
...
+                                printf("Info: fall back to CPU\n");
+                            } else {
+                                printf("Info: try to load next device: %s\n", device.c_str());
+                            }
+                        } else {
+                            noValidDevice = true;
+                        }
+                    } catch(const std::exception& e) {
+                        // select failed, there is no valid devices in deviceList
+                        noValidDevice = true;
```
[the remark this answers is not in the document]
Comment (tiger100256-hu) [reply inference-engine/src/multi_device/multi_device_exec_network.cpp]: for example ，the Customer try to load  "Auto:GPU,VPUX", and select GPU, GPU load failed, the selecting function continue, but VPUX don't support the network, the selecting  function will throw a exception

about exception message, I will add this message into load context

File: inference-engine/src/multi_device/multi_device_exec_network.cpp [hunk 5]
```diff
@@ -193,55 +193,123 @@ MultiDeviceExecutableNetwork::MultiDeviceExecutableNetwork(const std::string&
...
+                if (!isCPU) {
+                    _alreadyActualNetwork = true;
+                    _acceleratorPromise.set_value(executableNetwork);
+                } else {
+                    _cpuPromise.set_value(executableNetwork);
+                }
             } else {
-                executableNetwork = _core->LoadNetwork(network, device, deviceConfig);
+                if (!isCPU) {
+                    _acceleratorPromise.set_exception(eptr);
```
[the remark this answers is not in the document]
Comment (tiger100256-hu) [reply inference-engine/src/multi_device/multi_device_exec_network.cpp]: removed "set_exception",  just save error message

File: inference-engine/src/multi_device/multi_device_exec_network.cpp [hunk 6]
```diff
@@ -193,55 +193,123 @@ MultiDeviceExecutableNetwork::MultiDeviceExecutableNetwork(const std::string&
...
+                } else {
+                    _cpuPromise.set_exception(eptr);
+                }
             }
+            // the first device is already ready
+            std::call_once(_firstReadyOC, [this] () {
+                    _firstReadyPromise.set_value();
```
[the remark this answers is not in the document]
Comment (tiger100256-hu) [reply inference-engine/src/multi_device/multi_device_exec_network.cpp]: change the name from ”firstReadyOC“ to "firstLoadOC", the flag represent that there is load process finished, the process maybe failed

File: inference-engine/src/multi_device/multi_device_exec_network.cpp [hunk 7]
```diff
@@ -193,55 +193,123 @@ MultiDeviceExecutableNetwork::MultiDeviceExecutableNetwork(const std::string&
...
     }
-    if (_cpuFuture.valid() && _acceleratorFuture.valid()) {
-        try {
-            _networkFirstReady = _cpuFuture.get();
-        } catch (const std::exception& e) {
-            printf("Warning: load network to CPU failed: %s\n", e.what());
+    try {
+        if (_cpuFuture.valid() && _acceleratorFuture.valid()) {
+            try {
+                _networkFirstReady = _cpuFuture.get();
```
[the remark this answers is not in the document]
Comment (tiger100256-hu) [reply inference-engine/src/multi_device/multi_device_exec_network.cpp]: refactor the code and the code became sample

File: inference-engine/src/multi_device/multi_device_exec_network.cpp [hunk 8]
```diff
@@ -193,55 +193,123 @@ MultiDeviceExecutableNetwork::MultiDeviceExecutableNetwork(const std::string&
...
+            } catch (...) {
+                _networkActualNeeded = _acceleratorFuture.get();
+            }
+        } else if (_acceleratorFuture.valid()) {  // only accelerator is valid, like AUTO:GPU
             _networkActualNeeded = _acceleratorFuture.get();
+        } else if (_cpuFuture.valid()) {  // only CPU is valid, like AUTO:CPU
+            _networkActualNeeded = _cpuFuture.get();
+        } else {
+            IE_THROW() << "should not come here";
```
[the remark this answers is not in the document]
Comment (tiger100256-hu) [reply inference-engine/src/multi_device/multi_device_exec_network.cpp]: refactor the code and the remove this code

File: inference-engine/src/multi_device/multi_device_exec_network.cpp [hunk 9]
```diff
@@ -193,55 +193,123 @@ MultiDeviceExecutableNetwork::MultiDeviceExecutableNetwork(const std::string&
...
         }
-    } else if (_acceleratorFuture.valid()) {  // only accelerator is valid, like AUTO:GPU
-        _networkActualNeeded = _acceleratorFuture.get();
-    } else if (_cpuFuture.valid()) {  // only CPU is valid, like AUTO:CPU
-        _networkActualNeeded = _cpuFuture.get();
-    } else {
-        IE_THROW() << "No device task available";
+    } catch (...) {
+        IE_THROW() << "load all devices failed";
```
[the remark this answers is not in the document]
Comment (tiger100256-hu) [reply inference-engine/src/multi_device/multi_device_exec_network.cpp]: I wall add it

**Comment by tiger100256-hu** [comment]
I. refactor the code to make the code simple and add mock test case,
the refactor contains:
1. replace the variables of load network with loadcontext[2], make the code sample and reduce many uses of "if esle" 
2. replace  "do while" with recurrence function to simple the logic 
3. all the error message is save to errMessage of loadContext
the refactor logic is,
loadContext[ACTUALDEVICE] is always enabled,  when there is CPU and there are more than two devices, loadContext[CPU] is enabled

File: inference-engine/tests/unit/multi/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,49 @@
+# Copyright (C) 2018-2021 Intel Corporation
```
Comment (myshevts) [inline inference-engine/tests/unit/multi/CMakeLists.txt:1]: @nosovmik , you recently has implemenetd the mock-based tests for the caching, can you pls take a look on this PR from general sanity, e.g. whether we need the dedicated target, etc?

**Comment by myshevts** [comment]
generally I like the new code (for Unit tests) it just misses the comments, so it is hard to follow thru the test logic.
e.g. why we count the number of selections, but doesn't which device was actually selected?

File: inference-engine/src/multi_device/multi_device_exec_network.cpp [hunk 11]
```diff
@@ -158,106 +158,183 @@ MultiDeviceExecutableNetwork::MultiDeviceExecutableNetwork(const std::string&
...
-            IStreamsExecutor::ThreadBindingType::NONE});
-
-    for (auto& p : needLoadDevices) {
-        // initialize these containers firstly to avoid insert operation in threads
-        _idleWorkerRequests[p.deviceName];
-        _workerRequests[p.deviceName];
-        _inferPipelineTasksDeviceSpecific[p.deviceName] = NULL;
-        const auto device = p.deviceName;
-        const auto deviceConfig = p.config;
+    // initialize the flags of load network context
```
Comment (myshevts) [inline inference-engine/src/multi_device/multi_device_exec_network.cpp:161]: you may move that to the declaration of variables in the class, pls see below

**Review by myshevts — COMMENTED** [review]
@nosovmik , you are the only person who had expirience with the mock stuff recently, would you also have a look to the "mock" section of tis PR

File: inference-engine/src/multi_device/multi_device_exec_network.cpp [hunk 12]
```diff
@@ -159,109 +159,181 @@ MultiDeviceExecutableNetwork::MultiDeviceExecutableNetwork(const std::string&
...
+    for (int i = CONTEXTNUM - 1; i >= 0; i--) {
+        if (_loadContext[i].isEnabled && _loadContext[i].isAlready) {
+            return;
+        }
+    }
+
+    // the first loading is failed, wait for another loading
+    for (int i = CONTEXTNUM - 1; i >= 0; i--) {
+        if (_loadContext[i].isEnabled) {
+            _loadContext[i].future.wait();
```
Comment (myshevts) [inline inference-engine/src/multi_device/multi_device_exec_network.cpp:308]: not sure why this loop cannot be combined with the below loop... when the wait completed, we are already on the right index, so only need to check isAlready[i]

File: inference-engine/src/multi_device/multi_device_exec_network.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (tiger100256-hu) [reply inference-engine/src/multi_device/multi_device_exec_network.cpp]: "SelectDevice()" don't know if "deviceList" sorted, and deviceList maybe not all support the network. 
do you want "SelectDevice()" return a sorted devicelist, or something else

File: inference-engine/src/multi_device/multi_device_exec_network.cpp (hunk 11 above)
Comment (tiger100256-hu) [reply inference-engine/src/multi_device/multi_device_exec_network.cpp]: fix it

File: inference-engine/src/multi_device/multi_device_exec_network.hpp [hunk 13]
```diff
@@ -39,6 +39,26 @@ struct DeviceInformation {
     std::string defaultDeviceID;
 };
 
+struct AutoLoadContext {
+    std::atomic<bool> isEnabled;
```
[the remark this answers is not in the document]
Comment (tiger100256-hu) [reply inference-engine/src/multi_device/multi_device_exec_network.hpp]: fix it

File: inference-engine/src/multi_device/multi_device_exec_network.hpp [hunk 14]
```diff
@@ -39,6 +39,26 @@ struct DeviceInformation {
...
+    std::atomic<bool> isAlready;
+    std::atomic<bool> isLoadSuccess;
+    std::future<void> future;
+    std::promise<void> promise;
+    InferenceEngine::SoExecutableNetworkInternal executableNetwork;
+    DeviceInformation  deviceInfo;
+    std::vector<DeviceInformation> metaDevices;
```
[the remark this answers is not in the document]
Comment (tiger100256-hu) [reply inference-engine/src/multi_device/multi_device_exec_network.hpp:56]: networkPrecision and metaDevices are used only by _loadContext[Actual] for "selectDevice()",
_loadContext[CPU] didn't use and assign value for them.
add this two members into this Struct is because of 
1. metaDevices(devicelist) is changing in the loading process.  so need to save it 
2. add networkPrecision into struct can reduce this number of parameters of TryToLoadNetWork

File: inference-engine/src/multi_device/multi_device_exec_network.cpp (hunk 12 above)
Comment (tiger100256-hu) [reply inference-engine/src/multi_device/multi_device_exec_network.cpp:308]: fix it

File: inference-engine/src/multi_device/multi_device_exec_network.cpp [hunk 15]
```diff
@@ -158,106 +158,183 @@ MultiDeviceExecutableNetwork::MultiDeviceExecutableNetwork(const std::string&
...
-                executableNetwork = _core->LoadNetwork(modelPath, device, deviceConfig);
-            } else {
-                executableNetwork = _core->LoadNetwork(network, device, deviceConfig);
-            }
+        // so the executor can't be destroyed before finished the task,
+        // so use executor as a member of MultiDeviceExecutableNetwork.
+        _executor = InferenceEngine::ExecutorManager::getInstance()->getIdleCPUStreamsExecutor(
+                IStreamsExecutor::Config{"AutoDeviceAsyncLoad",
+                static_cast<int>(std::thread::hardware_concurrency()) /* max possible #streams*/,
+                1 /*single thread per stream*/,
```
[the remark this answers is not in the document]
Comment (tiger100256-hu) [reply inference-engine/src/multi_device/multi_device_exec_network.cpp]: fix it , set to 0

File: inference-engine/src/multi_device/multi_device_exec_network.cpp [hunk 16]
```diff
@@ -158,106 +158,183 @@ MultiDeviceExecutableNetwork::MultiDeviceExecutableNetwork(const std::string&
...
+                          GenerateWorkers(contextPtr->deviceInfo.deviceName, contextPtr->executableNetwork);
+                          //need lock
+                          {
+                             std::lock_guard<std::mutex> lock(_confMutex);
+                             _config.insert(contextPtr->deviceInfo.config.begin(),
+                                            contextPtr->deviceInfo.config.end());
+                          }
+                          contextPtr->isAlready = true;
+                      }
+                      contextPtr->promise.set_value();
```
[the remark this answers is not in the document]
Comment (tiger100256-hu) [reply inference-engine/src/multi_device/multi_device_exec_network.cpp:205]: the two async task need to set promise whether the load is successful or not, in the destroy function need to wait for the two task end, in some other functions also need to wait them

File: inference-engine/src/multi_device/multi_device_exec_network.cpp [hunk 17]
```diff
@@ -158,106 +158,183 @@ MultiDeviceExecutableNetwork::MultiDeviceExecutableNetwork(const std::string&
...
-    WaitFirstNetworkReady();
+    // select next candidate device
+    try {
+        context.deviceInfo = _multiPlugin->SelectDevice(deviceList, context.networkPrecision);
+    }
+    catch (const std::exception& e) {
+        return;
+    }
+
+    // if selec device is CPU, do not need to load CPU again, context[CPU] must have load CPU
```
[the remark this answers is not in the document]
Comment (tiger100256-hu) [reply inference-engine/src/multi_device/multi_device_exec_network.cpp]: yes, I agree with you, CPU accelerate should be as quick as possible , and it should be configured for latency at best.
Currently  the CPU use the same configures whether it's CPU accelerate or not  
I think it's another improve feature if need

File: inference-engine/src/multi_device/multi_device_exec_network.cpp [hunk 18]
```diff
@@ -158,106 +158,183 @@ MultiDeviceExecutableNetwork::MultiDeviceExecutableNetwork(const std::string&
...
-            if (!modelPath.empty()) {
-                executableNetwork = _core->LoadNetwork(modelPath, device, deviceConfig);
-            } else {
-                executableNetwork = _core->LoadNetwork(network, device, deviceConfig);
-            }
+        // so the executor can't be destroyed before finished the task,
+        // so use executor as a member of MultiDeviceExecutableNetwork.
+        _executor = InferenceEngine::ExecutorManager::getInstance()->getIdleCPUStreamsExecutor(
+                IStreamsExecutor::Config{"AutoDeviceAsyncLoad",
+                static_cast<int>(std::thread::hardware_concurrency()) /* max possible #streams*/,
```
[the remark this answers is not in the document]
Comment (tiger100256-hu) [reply inference-engine/src/multi_device/multi_device_exec_network.cpp:222]: I don't know whether it would affect the performance if set to 2, we can test it after our performance tests are setup totally, Currently just not modify it  from my point of view.

File: inference-engine/src/multi_device/multi_device_exec_network.hpp [hunk 19]
```diff
@@ -163,22 +189,21 @@ class MultiDeviceExecutableNetwork : public InferenceEngine::ExecutableNetworkTh
...
                                 const DeviceName& preferred_device);
+    void TryToLoadNetWork(AutoLoadContext& context,
+                          const std::string& modelPath,
+                          const InferenceEngine::CNNNetwork& network);
 
 private:
     std::shared_ptr<InferenceEngine::ICore>                             _core;
     InferenceEngine::IStreamsExecutor::Ptr                              _executor;
     MultiDeviceInferencePlugin*                                         _multiPlugin;
-    InferenceEngine::SoExecutableNetworkInternal                        _networkFirstReady;
```
Comment (nosovmik) [inline inference-engine/src/multi_device/multi_device_exec_network.hpp:171]: Recently merged PR #8389 uses this class member. Please rebase and fix build issues

File: inference-engine/src/multi_device/multi_device_exec_network.cpp [hunk 20]
```diff
@@ -495,11 +565,10 @@ InferenceEngine::Parameter MultiDeviceExecutableNetwork::GetConfig(const std::st
 InferenceEngine::Parameter MultiDeviceExecutableNetwork::GetMetric(const std::string &name) const {
     if (_workModeIsAUTO) {
         // fixme: should we wait actual device? meanwhile it will block inference, how to fix?
-        if (_alreadyActualNetwork) {
-            WaitActualNetworkReady();
-            return _networkActualNeeded->GetMetric(name);
+        if (_loadContext[ACTUALDEVICE].isAlready) {
```
Comment (nosovmik) [inline inference-engine/src/multi_device/multi_device_exec_network.cpp:568]: Is this covered by new tests?

File: inference-engine/tests/unit/multi/auto_select_device_failed_test.cpp [hunk 21]
```diff
@@ -0,0 +1,237 @@
...
+       // mock execNetwork can work
+       auto inferReqInternal = std::make_shared<MockIInferRequestInternal>();
+       ON_CALL(*mockIExeNet.get(), CreateInferRequest()).WillByDefault(Return(inferReqInternal));
+       IE_SET_METRIC(OPTIMAL_NUMBER_OF_INFER_REQUESTS, optimalNum, 2);
+       ON_CALL(*mockIExeNet.get(), GetMetric(StrEq(METRIC_KEY(OPTIMAL_NUMBER_OF_INFER_REQUESTS))))
+           .WillByDefault(Return(optimalNum));
+    }
+};
+
+TEST_P(AutoLoadFailedTest, LoadCNNetWork) {
```
Comment (nosovmik) [inline inference-engine/tests/unit/multi/auto_select_device_failed_test.cpp:123]: There is a lot of code changed and only this test is added. Is it enough?

**Review by nosovmik — CHANGES_REQUESTED** [review]
Please rebase and fix build issues on latest 'master'

File: inference-engine/src/multi_device/multi_device_exec_network.hpp (hunk 19 above)
Comment (tiger100256-hu) [reply inference-engine/src/multi_device/multi_device_exec_network.hpp:171]: rebased the code

File: inference-engine/tests/unit/multi/CMakeLists.txt [hunk 22]
```diff
@@ -0,0 +1,36 @@
...
+# SPDX-License-Identifier: Apache-2.0
+#
+
+set(TARGET_NAME ieMultiPluginUnitTests)
```
[the remark this answers is not in the document]
Comment (tiger100256-hu) [reply inference-engine/tests/unit/multi/CMakeLists.txt:5]: enabled this test in  '.ci/azure'

File: inference-engine/tests/unit/multi/CMakeLists.txt [hunk 23]
```diff
@@ -0,0 +1,36 @@
...
+
+set(CI_BUILD_NUMBER "unittest")
```
[the remark this answers is not in the document]
Comment (tiger100256-hu) [reply inference-engine/tests/unit/multi/CMakeLists.txt:7]: the plugin code need this variable to compile

File: inference-engine/src/multi_device/multi_device_exec_network.cpp [hunk 24]
```diff
@@ -159,109 +159,178 @@ MultiDeviceExecutableNetwork::MultiDeviceExecutableNetwork(const std::string&
...
+
+    // loadContext[ACTUALDEVICE] is always enabled,
+    // when there is CPU and there are more than two devices, loadContext[CPU] is enabled
+    _loadContext[ACTUALDEVICE].isEnabled = true;
+    _loadContext[ACTUALDEVICE].networkPrecision = GetNetworkPrecision(network);
+    _loadContext[ACTUALDEVICE].metaDevices = metaDevices;
+    _loadContext[ACTUALDEVICE].deviceInfo = _multiPlugin->SelectDevice(metaDevices, _loadContext[ACTUALDEVICE].networkPrecision);
+    bool isActualDevCPU =
+        _loadContext[ACTUALDEVICE].deviceInfo.deviceName.find("CPU") != std::string::npos;
+    // if Acutal device is CPU, diabled _loadContext[CPU], only use _loadContext[ACTUALDEVICE]
```
[the remark this answers is not in the document]
Comment (tiger100256-hu) [reply inference-engine/src/multi_device/multi_device_exec_network.cpp]: fix it

File: inference-engine/src/multi_device/multi_device_exec_network.cpp [hunk 25]
```diff
@@ -474,6 +543,7 @@ void MultiDeviceExecutableNetwork::SetConfig(const std::map<std::string, Inferen
 }
 
 InferenceEngine::Parameter MultiDeviceExecutableNetwork::GetConfig(const std::string &name) const {
+    std::lock_guard<std::mutex> confLock(_confMutex);
```
[the remark this answers is not in the document]
Comment (tiger100256-hu) [reply inference-engine/src/multi_device/multi_device_exec_network.cpp]: fix it

File: inference-engine/src/multi_device/multi_device_plugin.hpp [hunk 26]
```diff
@@ -33,11 +40,11 @@ class MultiDeviceInferencePlugin : public InferenceEngine::IInferencePlugin {
     InferenceEngine::Parameter GetMetric(const std::string& name,
                                          const std::map<std::string, InferenceEngine::Parameter>& options) const override;
 
-    std::vector<MultiDevicePlugin::DeviceInformation> ParseMetaDevices(const std::string & devicesRequestsCfg,
+    MOCKTESTMACRO std::vector<MultiDevicePlugin::DeviceInformation> ParseMetaDevices(const std::string & devicesRequestsCfg,
```
[the remark this answers is not in the document]
Comment (tiger100256-hu) [reply inference-engine/src/multi_device/multi_device_plugin.hpp:43]: yes, I also think it's ugly, but it's very easy to test many functions in the plugin if need, the logic in auto plugin will be more and more complex. if all functions can be tested by gmock,  that will make our plugin code better. Maybe there are other way to do it

File: inference-engine/src/multi_device/multi_device_exec_network.cpp [hunk 27]
```diff
@@ -159,109 +159,178 @@ MultiDeviceExecutableNetwork::MultiDeviceExecutableNetwork(const std::string&
...
+        _executor = InferenceEngine::ExecutorManager::getInstance()->getIdleCPUStreamsExecutor(
+                IStreamsExecutor::Config{"AutoDeviceAsyncLoad",
+                static_cast<int>(std::thread::hardware_concurrency()) /* max possible #streams*/,
+                0 /*default threads per stream, workaround for ticket 62376*/,
+                IStreamsExecutor::ThreadBindingType::NONE});
+        for (auto&& device : metaDevices) {
+            // initialize containers before run async task
+            _idleWorkerRequests[device.deviceName];
+            _workerRequests[device.deviceName];
+            _inferPipelineTasksDeviceSpecific[device.deviceName] = NULL;
```
[the remark this answers is not in the document]
Comment (tiger100256-hu) [reply inference-engine/src/multi_device/multi_device_exec_network.cpp]: fix it

File: inference-engine/src/multi_device/multi_device_exec_network.cpp (hunk 20 above)
Comment (tiger100256-hu) [reply inference-engine/src/multi_device/multi_device_exec_network.cpp:582]: except this mock test, there is also function test, function test should cover this API, but not 100% sure.
we will check code coverage

**Review by myshevts — CHANGES_REQUESTED** [review]

File: inference-engine/src/multi_device/multi_device_exec_network.cpp [hunk 28]
```diff
@@ -159,109 +159,181 @@ MultiDeviceExecutableNetwork::MultiDeviceExecutableNetwork(const std::string&
...
+        _executor->run(_loadContext[ACTUALDEVICE].task);
+    } else {
+        // only one device need to load network, do not need to load it async
+        _loadContext[ACTUALDEVICE].task();
+    }
 
-            GenerateWorkers(device, executableNetwork);
+    WaitFirstNetworkReady();
+}
+void MultiDeviceExecutableNetwork::TryToLoadNetWork(AutoLoadContext& context,
```
[the remark this answers is not in the document]
Comment (myshevts) [reply inference-engine/src/multi_device/multi_device_exec_network.cpp:240]: this is somewhat minor if you managed to implement the unit tests for the SelectDevice without making the methods static, but I wish I publish this comment much earlier, that would probably simplify you the developemnt of the units tests

File: inference-engine/src/multi_device/multi_device_exec_network.cpp [hunk 29]
```diff
@@ -159,109 +159,181 @@ MultiDeviceExecutableNetwork::MultiDeviceExecutableNetwork(const std::string&
...
     // Maybe different API will call this function, so add call once here
     // for every MultiDeviceExecutableNetwork instance
-    OV_ITT_SCOPED_TASK(itt::domains::MULTIPlugin, "MultiDeviceExecutableNetwork::WaitActualNetworkReady");
-    std::call_once(_oc, [&] () {
-            if (_acceleratorFuture.valid()) {
-                _networkActualNeeded = _acceleratorFuture.get();
-            }
+    std::call_once(_oc, [this] () {
+               if (_loadContext[ACTUALDEVICE].future.valid()) {
+                   _loadContext[ACTUALDEVICE].future.get();
```
[the remark this answers is not in the document]
Comment (tiger100256-hu) [reply inference-engine/src/multi_device/multi_device_exec_network.cpp:326]: if an exception is not handle, it will cause the application exit.
if promise set_exception,  it will cause future.get() to re-thrown exception
in this code, we handled the failed load exception, but not set exception, so it will not re-thrown

File: inference-engine/src/multi_device/multi_device_exec_network.cpp [hunk 30]
```diff
@@ -329,7 +400,8 @@ void MultiDeviceExecutableNetwork::run(Task inferPipelineTask) {
 
 MultiDeviceExecutableNetwork::~MultiDeviceExecutableNetwork() {
     // this is necessary to guarantee member destroyed after getting future
-    if (_workModeIsAUTO) {
+    if (_workModeIsAUTO && _loadContext[CPU].isEnabled) {
+        _loadContext[CPU].future.get();
```
[the remark this answers is not in the document]
Comment (tiger100256-hu) [reply inference-engine/src/multi_device/multi_device_exec_network.cpp:401]: Yes, we didn't use future.valid(), so it's same to use wait and get here, 
about exception in load network has be handled, it will not throw exception again.
A remind, the loading process contain "loadNetwork" and "GenerateWorkers"， if "GenerateWorkers" thrown exception, it also will cause the app exit

File: inference-engine/src/multi_device/multi_device_exec_network.cpp (hunk 28 above)
[the remark this answers is not in the document]
Comment (tiger100256-hu) [reply inference-engine/src/multi_device/multi_device_exec_network.cpp:240]: Yes, from my opinion, I want to find a way to test all the function in the plugin, I maybe could change some function to static, but can't make all function to static. Currently I add some mock macro into head file, it's ugly, but it really help me to test the functions, I really concern is about if this way can work, until now, it work and haven't meet regression

File: inference-engine/src/multi_device/multi_device_exec_network.cpp (hunk 29 above)
[the remark this answers is not in the document]
Comment (myshevts) [reply inference-engine/src/multi_device/multi_device_exec_network.cpp:326]: in the destructor it is may be good not to re-throw anything , as if there is another exception being propagated (e.g. the actual reason of the app doing the exit) the terminate will be called.

**Review by myshevts — APPROVED** [review]
